### ISSUE TITLE
Add biology words to ML model

### ISSUE THREAD

**Issue by bruffridge** [nasa-petal/PeTaL-labeller#13]
PAHT: Talking to Shruti and Herb, we have some issues. what we are doing now is taking a model from Bert and the corresponding tokenizer. Both of which were trained on the English language (OSCAR) and amazon reviews, text to friends, angry blogs etc. We are taking that and applying it to Biology papers. A tokenizer converts words in English to numbers for the machine to understand. The model is used to predict understand how the "tokens/words" fit together. This tells it how to classify stuff.

Problem: We have words in biology that bert was probably never trained to understand. This would definitely lower the prediction accuracy. Bert is currently predicting using it's own word database (no biology words)

Solutions

1. Mine papers from biology. Create a clean nice dataset of sentences using biology words, basically our own Corpus for biology (kind of like https://traces1.inria.fr/oscar/).

    (Option a) We take bert, figure out how to add our own tokens (new biology words) and train it to understand how the new tokens relate to each other from our corpus. This should be the first choice.
    (Option b) We train our own bert model and tokenizer based on our corpus. Hardest, most computational resources. We would be reinventing some poorly documented wheel.

Having a corpus for biology words would hypothetically reduce our need of more data from MTurk. Making a corpus from MTurk is way more costly than mining data from a paper and formatting it correctly. Make sure we dont have weird symbols.

**Comment by bruffridge** [comment]
HERB: Would using this help? https://github.com/dmis-lab/biobert

**Comment by bruffridge** [comment]
ALEX: What about PubMedBERT? https://huggingface.co/microsoft/BiomedNLP-PubMedBERT-base-uncased-abstract

**Comment by bruffridge** [comment]
HERB: Some of these are newly posted, I think. Not sure how applicable they are https://paperswithcode.com/lib/allennlp

**Comment by bruffridge** [comment]
> HERB: Would using this help? https://github.com/dmis-lab/biobert

SHRUTI: I thought about that too but it doesn’t seem like they have a specific tokenizer for it.
But they have a vocabulary that can be added that we could possibly utilize?

![Image from iOS](https://user-images.githubusercontent.com/1322063/111473237-3347e980-8701-11eb-8449-b8375b590a11.png)

**Comment by pjuangph** [comment]
We should train our own. I would also like to see if you can simply run a few abstracts to see if we are missing words in the tokenizer and if so which words. Give an example. 

If we can automate the downloading of all the abstracts and papers, we can train our own tokenizer on them. There is no guarantee biobert has everything we need.

**Comment by bruffridge** [comment]
Shruti confirmed that the BERT tokenizer did not identify many of the scientific terms in the abstracts. She recommends trying a different tokenizer or adding our own dictionary to the tokenizer. @shrutix correct me if I got any of the details wrong or missed anything.

**Comment by pjuangph** [comment]
Good job @shrutix more than likely we have to train our own tokenizer and our own bert. 
I think shruti is right to try biobert. I would also like to know if biobert is any better.

**Comment by shrutix** [comment]
From the BioBERT paper, on tokenization:

> For tokenization, BioBERT uses WordPiece tokenization (Wu et al., 2016), which mitigates the out-of-vocabulary issue. With WordPiece tokenization, any new words can be represented by fre- quent subwords (e.g. Immunoglobulin 1⁄4> I ##mm ##uno ##g ##lo ##bul ##in). We found that using cased vocabulary (not lower- casing) results in slightly better performances in downstream tasks. Although we could have constructed new WordPiece vocabulary based on biomedical corpora, we used the original vocabulary of BERTBASE for the following reasons: (i) compatibility of BioBERT with BERT, which allows BERT pre-trained on general domain cor- pora to be re-used, and makes it easier to interchangeably use exist- ing models based on BERT and BioBERT and (ii) any new words may still be represented and fine-tuned for the biomedical domain using the original WordPiece vocabulary of BERT.

**Comment by hschilling** [comment]
> From the BioBERT paper, on tokenization:
> 
> > For tokenization, BioBERT uses WordPiece tokenization (Wu et al., 2016), which mitigates the out-of-vocabulary issue. With WordPiece tokenization, any new words can be represented by fre- quent subwords (e.g. Immunoglobulin 1⁄4> I ##mm ##uno ##g ##lo ##bul ##in). We found that using cased vocabulary (not lower- casing) results in slightly better performances in downstream tasks. Although we could have constructed new WordPiece vocabulary based on biomedical corpora, we used the original vocabulary of BERTBASE for the following reasons: (i) compatibility of BioBERT with BERT, which allows BERT pre-trained on general domain cor- pora to be re-used, and makes it easier to interchangeably use exist- ing models based on BERT and BioBERT and (ii) any new words may still be represented and fine-tuned for the biomedical domain using the original WordPiece vocabulary of BERT.

Great info, @shrutix  !

**Comment by shrutix** [comment]
Thanks! Tentative plan with the vocabulary:

> not sure if this will work accurately, since we’re essentially training BERT on embeddings/words it may have not been pretrained with through WordPiece tokenization, but I think it’s still worth it to try.
Essentially, scientifically related terms such as defoliation or ectomycorrhizal are being tokenized as 'def', '##oli', '##ation and 'ec', '##tom', '##y', '##cor', '##rh', '##iza', '##l'  instead of the entire word itself. I want to try adding words like this directly into the bert tokenizer (manually at this point) to see if it can correct itself and analyze the words and their context correctly. I’m going to go through the abstract dataset I have to find any words like this that are being incorrectly tokenized and add them to the tokenizer. Also collaborating with Alex and Colleen to compile a list of any other biomimicry/biology-related terms like these that may appear in any new abstracts that we will use in the future. Starting with https://biologydictionary.net/complete-list-biology-terms/

**Comment by bruffridge** [comment]
Using SciBERT improved tokenization. https://www.semanticscholar.org/paper/SciBERT%3A-A-Pretrained-Language-Model-for-Scientific-Beltagy-Lo/5e98fe2163640da8ab9695b9ee9c433bb30f5353

**Comment by pjuangph** [comment]
Resolved in latest commit 3992cee949e74e105239e92bb9592d04d017661c
